crates/walrus-sdk/src/client.rs

```diff
@@ -1549,6 +1551,46 @@ impl<T> WalrusNodeClient<T> {
         blobs: &[(VerifiedBlobMetadataWithId, Arc<Vec<SliverPair>>)],
         event_sender: tokio::sync::mpsc::Sender<UploaderEvent>,
         tail_handling: TailHandling,
+        cancellation: Option<CancellationToken>,
+    ) -> ClientResult<RunOutput<Vec<BlobId>, StoreError>> {
+        self.distributed_upload_without_confirmation_inner(
+            blobs,
+            event_sender,
+            tail_handling,
+            cancellation,
+            None,
+        )
+        .await
+    }
+
+    /// Targets only the provided node indices.
+    /// This is useful when retrying uploads without hitting already-successful nodes.
+    #[allow(dead_code)]
+    pub(crate) async fn distributed_upload_without_confirmation_for_nodes(
+        &self,
+        blobs: &[(VerifiedBlobMetadataWithId, Arc<Vec<SliverPair>>)],
+        event_sender: tokio::sync::mpsc::Sender<UploaderEvent>,
+        tail_handling: TailHandling,
+        cancellation: Option<CancellationToken>,
+        node_indices: &[NodeIndex],
+    ) -> ClientResult<RunOutput<Vec<BlobId>, StoreError>> {
+        self.distributed_upload_without_confirmation_inner(
+            blobs,
+            event_sender,
+            tail_handling,
+            cancellation,
+            Some(node_indices),
+        )
+        .await
+    }
+
+    async fn distributed_upload_without_confirmation_inner(
+        &self,
+        blobs: &[(VerifiedBlobMetadataWithId, Arc<Vec<SliverPair>>)],
+        event_sender: tokio::sync::mpsc::Sender<UploaderEvent>,
+        tail_handling: TailHandling,
+        cancellation: Option<CancellationToken>,
+        target_nodes: Option<&[NodeIndex]>,
```

Comment:
question: this index is specific to one committee in one epoch, is that correct? What if the committee got inside this call is changed from what it was returned? Say a new epoch just started?

Reply:
yes node index is tied to a specific epoch, i've now ensured that we pass that particular epoch that is used to derive the index and verify that committee epoch hasn't moved forward and surface an error which callers can handle by refreshing the committee.